### PR TITLE
[scope] improve 'scope' and 'return' inference for delegates

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

File: test/fail_compilation/retscope.d [hunk 1]
```diff
@@ -235,13 +235,13 @@ void* funretscope(scope dg_t ptr) @safe
...
-fail_compilation/retscope.d(249): Error: cannot implicitly convert expression (__lambda1) of type void* delegate() pure nothrow @nogc return @safe to void* delegate() @safe
-fail_compilation/retscope.d(250): Error: cannot implicitly convert expression (__lambda2) of type void* delegate() pure nothrow @nogc return @safe to void* delegate() @safe
-fail_compilation/retscope.d(250): Error: cannot implicitly convert expression (__lambda2) of type void* delegate() pure nothrow @nogc return @safe to void* delegate() @safe
 ---
 */
 
+
+
+
+
```
Comment (mathias-lang-sociomantic) [inline test/fail_compilation/retscope.d]: Does that mean those are implicitly convertible now ?

Comment (WalterBright) [inline test/fail_compilation/retscope.d]: yes

Comment (Geod24) [inline test/fail_compilation/retscope.d]: No you didn't. The following snippet (rightfully) doesn't compile with master, but does with your P.R.:

```D
struct Bar { int i; }

alias FunDG = Bar* delegate () @safe;

void main () @safe
{
    Bar* b = getBar();
}

Bar* getBar () @safe
{
    Bar b;
    scope FunDG x = () return { return &b; };
    return forward(x);
}

Bar* forward (scope FunDG escaper) @safe
{
    return escaper();
}
```

Comment (WalterBright) [inline test/fail_compilation/retscope.d]: You're right. Will fix.

Comment (WalterBright) [inline test/fail_compilation/retscope.d]: If `FunDG` itself is declared as `delegate () scope` then the error is diagnosed. So the trouble is adding `scope` on later. Will investigate.

**Comment by WalterBright** [comment]
> Wrong base branch, must be stable.

I did a separate one for stable:
https://github.com/dlang/dmd/pull/6497

**Comment by WalterBright** [comment]
This needs to be reopened. This fix is much more complete, and is needed.

**Comment by WalterBright** [comment]
Well, maybe I should just redo it as a separate PR, to avoid confusion. I'll think about it.

**Comment by WalterBright** [comment]
@MartinNowak doesn't need to sequence the merge. He can just merge it - I'll rebase this one when he does.

**Comment by MartinNowak** [comment]
Please rebase, stable change already in master since https://github.com/dlang/dmd/pull/6498.

**Comment by WalterBright** [comment]
Rebase done.

**Comment by WalterBright** [comment]
Currently blocked by https://github.com/dlang/druntime/pull/1753

**Comment by MartinNowak** [comment]
Looking at https://github.com/dlang/druntime/pull/1753, this will be fairly unusable in practice. People have added scope to delegate parameters basically everywhere to avoid GC closure allocations, now requiring to pass a delegate that actually is scope will break lots of code. Let's please do this as a deprecation.

**Comment by WalterBright** [comment]
@MartinNowak the changes to druntime are necessary because druntime is now compiled with -dip1000. This does not affect non-dip1000 builds.

As the numerous druntime and phobos PRs I've submitted attest, compiling with -dip1000 will need changes here and there. There is no other way I can think of - and that's why it's behind a switch. The transition period will be a long one.

It's turning out to be a bit tricky to get code to compile both with and without -dip1000, but it's worth the effort.

**Comment by mathias-lang-sociomantic** [comment]
> People have added scope to delegate parameters basically everywhere to avoid GC closure allocations, now requiring to pass a delegate that actually is scope will break lots of code.

I'm not sure I follow your comment. The issue in the linked P.R. seems to be about a `scope` return, as opposed to a `scope` delegate (e.g. the missing `scope` on the RHS is the issue, not the one on the LHS).

**Comment by WalterBright** [comment]
The `scope` attribute, whether it is on the left or right side, refers to the `this` parameter, i.e. the `.funcptr` as that is the value that should not escape.

**Comment by andralex** [comment]
Auto-merge toggled on

**Comment by MartinNowak** [comment]
> As the numerous druntime and phobos PRs I've submitted attest, compiling with -dip1000 will need changes here and there. There is no other way I can think of - and that's why it's behind a switch. The transition period will be a long one.
> 
> It's turning out to be a bit tricky to get code to compile both with and without -dip1000, but it's worth the effort.

Let me be a bit clearer on my comment (https://github.com/dlang/dmd/pull/6495#issuecomment-277932866).
Of course everything is behind the `-dip1000` switch for now. But the change in _this_ PR will affect much more people than most of the others, b/c scope for delegate parameters were one of the few use-cases of scope prior to DIP1000.
Requiring people to rewrite all their code and dependencies before being able to even test `-dip1000` will hamper adoption. So for this scope check it would be particularly helpful if it were a deprecation instead of an error (both behind -dip1000 of course).

**Comment by WalterBright** [comment]
> Requiring people to rewrite all their code and dependencies before being able to even test -dip1000 will hamper adoption.

I don't see another way. It's taking me a while to work through Phobos on this, and sometimes the rat's nest of circular dependencies in there make things difficult. But the ends are worth it. (Heck, we're still working on making Phobos @safe!) It will be quite a while until we can make -dip1000 the default.

**Comment by MartinNowak** [comment]
> I don't see another way.

Just make it a deprecation instead of an error, so that things can be adopted incrementally.

**Comment by MartinNowak** [comment]
Remaining unexpected test failure/diff.
```
fail_compilation/retscope.d(1867): Error: retscope.foo22 called with argument types (char[]) matches both:
fail_compilation/retscope.d(1860):     retscope.foo22!().foo22(ref char[] s)
and:
fail_compilation/retscope.d(1894):     retscope.foo22!().foo22(ref char[] s)
```

**Comment by WalterBright** [comment]
Blocked by https://github.com/dlang/druntime/pull/1762

**Pull request merged**